In [355]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
dataset = load_boston()
print(dataset.feature_names)
print(dataset.data)

['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']
[[6.3200e-03 1.8000e+01 2.3100e+00 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [2.7310e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [2.7290e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [6.0760e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0959e-01 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [4.7410e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 7.8800e+00]]


/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (

In [356]:
print(dataset.target)

[24.  21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 15.  18.9 21.7 20.4
 18.2 19.9 23.1 17.5 20.2 18.2 13.6 19.6 15.2 14.5 15.6 13.9 16.6 14.8
 18.4 21.  12.7 14.5 13.2 13.1 13.5 18.9 20.  21.  24.7 30.8 34.9 26.6
 25.3 24.7 21.2 19.3 20.  16.6 14.4 19.4 19.7 20.5 25.  23.4 18.9 35.4
 24.7 31.6 23.3 19.6 18.7 16.  22.2 25.  33.  23.5 19.4 22.  17.4 20.9
 24.2 21.7 22.8 23.4 24.1 21.4 20.  20.8 21.2 20.3 28.  23.9 24.8 22.9
 23.9 26.6 22.5 22.2 23.6 28.7 22.6 22.  22.9 25.  20.6 28.4 21.4 38.7
 43.8 33.2 27.5 26.5 18.6 19.3 20.1 19.5 19.5 20.4 19.8 19.4 21.7 22.8
 18.8 18.7 18.5 18.3 21.2 19.2 20.4 19.3 22.  20.3 20.5 17.3 18.8 21.4
 15.7 16.2 18.  14.3 19.2 19.6 23.  18.4 15.6 18.1 17.4 17.1 13.3 17.8
 14.  14.4 13.4 15.6 11.8 13.8 15.6 14.6 17.8 15.4 21.5 19.6 15.3 19.4
 17.  15.6 13.1 41.3 24.3 23.3 27.  50.  50.  50.  22.7 25.  50.  23.8
 23.8 22.3 17.4 19.1 23.1 23.6 22.6 29.4 23.2 24.6 29.9 37.2 39.8 36.2
 37.9 32.5 26.4 29.6 50.  32.  29.8 34.9 37.  30.5 36.4 31.1 29.1 50.
 33.3 3

## *Housing Prices Above*
## *Loading into Pandas Dataframe Below*

In [357]:
df = pd.DataFrame(dataset.data, columns=dataset.feature_names)
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

## *Figure 6.2 Above*
## *Adding Housing Prices to Dataframe Below*

In [358]:
df['MEDV'] = dataset.target
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

## *Figure 6.3 Above*
## *Data Cleansing Below*

In [359]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(14)
memory usage: 55.5 KB


## *Above: Scikit-learn only works with fields that are numeric, you need to encode string values into numeric values. Fortunately, the dataset contains all numerical values, and so no encoding is necessary.*

## *Below: We need to check to see if there are any missing values. To do so, use the isnull() function:*

In [360]:
print(df.isnull().sum())

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64


## *The dataset is good, as it does not have any missing values above*


# *Feature Selection*

## *There are 13 features in the dataset, we do not want to use all of these features for training our model, because not all of them are relevant.  Instead, we want to choose those features that directly influence the result (that is, prices of houses) to train the model. For this, we can use the corr() function.* 
## *The corr() function computes the pairwise correlation of columns:* 

In [361]:
corr = df.corr()
print(corr)

             CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.200469  0.406583 -0.055892  0.420972 -0.219247  0.352734   
ZN      -0.200469  1.000000 -0.533828 -0.042697 -0.516604  0.311991 -0.569537   
INDUS    0.406583 -0.533828  1.000000  0.062938  0.763651 -0.391676  0.644779   
CHAS    -0.055892 -0.042697  0.062938  1.000000  0.091203  0.091251  0.086518   
NOX      0.420972 -0.516604  0.763651  0.091203  1.000000 -0.302188  0.731470   
RM      -0.219247  0.311991 -0.391676  0.091251 -0.302188  1.000000 -0.240265   
AGE      0.352734 -0.569537  0.644779  0.086518  0.731470 -0.240265  1.000000   
DIS     -0.379670  0.664408 -0.708027 -0.099176 -0.769230  0.205246 -0.747881   
RAD      0.625505 -0.311948  0.595129 -0.007368  0.611441 -0.209847  0.456022   
TAX      0.582764 -0.314563  0.720760 -0.035587  0.668023 -0.292048  0.506456   
PTRATIO  0.289946 -0.391679  0.383248 -0.121515  0.188933 -0.355501  0.261515   
B       -0.385064  0.175520 

## *A perfect negative correlation is represented by the value –1.00: a 0.00 indicates no correlation and a +1.00 indicates a perfect positive correlation.*

## *From the MEDV column above in the output, you can see that the RM and LSTAT features have high correlation factors (positive and negative correlations) with the MEDV.* 

## *Below, instead of visually finding the high correlation variables we'll find the top 3 variables (MEDV is the 3rd but not counted) programmatically:*

In [362]:
#---get the top 4 features that has the highest correlation---
print(df.corr().abs().nlargest(3, 'MEDV').index)
#---print the top 4 correlation values---
print(df.corr().abs().nlargest(3, 'MEDV').values[:,13])

Index(['MEDV', 'LSTAT', 'RM'], dtype='object')
[1.         0.73766273 0.69535995]


## *Above you can see the 2 high correlation variables are LSTAT and RM.  MEDV also appears but can't be counted*

<div class="alert alert-block alert-info">
Here is where we begin training the model
</div>

## *Training the Model Below*
## *We can now train the model. First, create two DataFrames: x and Y. The x DataFrame will contain the combination of the LSTAT and RM features, while the Y DataFrame will contain the MEDV label:*

In [363]:
x = pd.DataFrame(np.c_[df['LSTAT'], df['RM']], columns = ['LSTAT','RM'])
Y = df['MEDV']

## *Now we'll print x to make sure we have the correct 2 columns and data*

In [364]:
print(x)

     LSTAT     RM
0     4.98  6.575
1     9.14  6.421
2     4.03  7.185
3     2.94  6.998
4     5.33  7.147
..     ...    ...
501   9.67  6.593
502   9.08  6.120
503   5.64  6.976
504   6.48  6.794
505   7.88  6.030

[506 rows x 2 columns]


## *Now we'll print Y just to check that MEDV is correct*

In [365]:
print(Y)

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
       ... 
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: MEDV, Length: 506, dtype: float64


<div class="alert alert-block alert-info">
Here is where we split the model
</div>

## *Below we will split the dataset into 70 percent for training and 30 percent for testing:*

In [366]:
from sklearn.model_selection import train_test_split
x_train, x_test, Y_train, Y_test = train_test_split(x, Y, test_size = 0.3, random_state=5)

## *After the split, let’s print out the shape of the training sets below*

In [367]:
print(x_train.shape)
print(Y_train.shape)

(354, 2)
(354,)


## *This means that the x training set now has 354 rows and 2 columns, while the Y training set (which contains the label) has 354 rows and 1 column.*

## *Let’s also print out the testing set:*

In [368]:
print(x_test.shape)
print(Y_test.shape)

(152, 2)
(152,)


## *This time, the testing set has 152 rows shown above*

## *We are now ready to begin the training. As you learned from the previous chapter, you can use the LinearRegression class to perform linear regression. In this case, we will use it to train our model:*

In [369]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x_train, Y_train)

LinearRegression()

## *With the model trained, we can obtain the intercept as well as the coefficients of the features:*

In [370]:
print(model.intercept_)
print(model.coef_)

0.3843793678034899
[-0.65957972  4.83197581]


## *We can use the model to make a prediction against the actual data for the house price when LSTAT is 4.98 and RM is 6.575 (data from row 1):*

In [371]:
print(model.predict([[4.98,6.575]]))

[28.86991336]


/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


## *For comparison we can see how the output (MEDV 28.8699) compares to row 1 MEDV from the data below (MEDV 24.0)* 

In [372]:
df['MEDV'] = dataset.target
df.head(1)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE   DIS  RAD    TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.09  1.0  296.0     15.3   

       B  LSTAT  MEDV  
0  396.9   4.98  24.0

## *The R-Squared method lets us know how close the test data fits the regression line. A value of 1.0 means a perfect fit. We aim for a value of R-Squared that is close to 1:*

In [373]:
print('R-Squared: %.4f' % model.score(x_test, Y_test))

R-Squared: 0.6162


## *The R-Squared value is 0.6162*

## *Now we'll try splitting the test and training up 80% training and 20% testing instead of the 70/30 ratio used earlier*

In [375]:
from sklearn.model_selection import train_test_split
x_train, x_test, Y_train, Y_test = train_test_split(x, Y, test_size = 0.2, random_state=5)

## *Below we can see 404 rows are used for training with the 80/20 ratio*

In [376]:
print(x_train.shape)
print(Y_train.shape)

(404, 2)
(404,)


## *Below we can see 102 rows are used for testing with the 80/20 ratio*

In [377]:
print(x_test.shape)
print(Y_test.shape)

(102, 2)
(102,)


In [378]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x_train, Y_train)

LinearRegression()

## *Now we run the same LSTAT and RM inputs from data row 1 to see comparison to actual MEDV from the data (24.0)*

In [380]:
print(model.predict([[4.98,6.575]]))

[29.33966548]


/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


## *You can see above that the MEDV prediction is 29.33966 which is further away from the actual value from the data of 24.0 compared to the first prediction of 28.8699*
## *So we can see that the 80/20 ratio hasn't been as accurate as the 70/30 ratio*
## *Below is the R-Squared value for the 80/20 ratio.  0.6629 is closer to 1.0 than the 70/30 ratio R-Squared value of 0.6162.  So the 80/20 ratio MEDV prediction was less accurate but the R-Squared value was more accurate than the 70/30 ratio.*

In [381]:
print('R-Squared: %.4f' % model.score(x_test, Y_test))

R-Squared: 0.6629


## *Now we'll try splitting the test and training up 85% training and 15% testing*

In [382]:
from sklearn.model_selection import train_test_split
x_train, x_test, Y_train, Y_test = train_test_split(x, Y, test_size = 0.15, random_state=5)

## *Below we can see 430 rows are used for training with the 85/15 ratio*

In [383]:
print(x_train.shape)
print(Y_train.shape)

(430, 2)
(430,)


## *We can also see below that 76 rows are used for testing with the 85/15 ratio*

In [384]:
print(x_test.shape)
print(Y_test.shape)

(76, 2)
(76,)


In [385]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x_train, Y_train)

LinearRegression()

## *Here we run the MEDV prediction again with the 85/15 training testing split*

In [386]:
print(model.predict([[4.98,6.575]]))

[29.31555873]


/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


## *We can see above that 85/15 ratio MEDV prediction (29.31555) is closer to the actual data value from row 1 (24.0) than the 80/20 split prediction (29.33966) but still less accurate than the 70/30 split prediction (28.8699)*

## *Below the R-Squared value for the 85/15 ratio is 0.6389 which is closer to 1.0 than the 70/30 ratio R-Squared value of 0.6162, but less accurate than the R-Squared value of 0.6629 for the 80/20 ratio.*  
## *So the 85/15 ratio MEDV prediction was less accurate than the 70/30 ratio prediction but the R-Squared value was more accurate than the 70/30 ratio.*

In [387]:
print('R-Squared: %.4f' % model.score(x_test, Y_test))

R-Squared: 0.6389
